dnn_rotate/nodes/rotate.py

```diff
@@ -0,0 +1,154 @@
+        rospy.loginfo("Waiting for move service to be available")
+        if not self.move.wait_for_server(rospy.Duration(10.0)):
+           rospy.logerr("Move service not available")
+           return
+
+        # Create a proxy object fot the trigger service we call
+        rospy.loginfo("Waiting for detect service to be available")
+        try:
+            rospy.wait_for_service("/dnn_detect/detect", 10)
+            self.trigger = rospy.ServiceProxy("/dnn_detect/detect",
+                                              dnn_detect.srv.Detect)
+            rospy.loginfo("Rotate node ready")
+            self.ok = True
+        except:
+            rospy.logerr("detect service is not available") 
+
+    # This is called when we receive a rotate service call
+    def service_callback(self, rotate_req):
+        target = rotate_req.object.data
+        print("Received service call: rotate %s" % target)
+
+        # Create a response to our service
+        response = dnn_rotate.srv.StringResponse()
+
+        # Make a detection trigger service call
+        detect_req = dnn_detect.srv.DetectRequest()
+        detect_resp = self.trigger(detect_req)
+
+        # Look for matching objects
+        best_target = None
+        best_confidence = 0.0
+        for object in detect_resp.result.objects:
+            if object.class_name == target:
+                found_target = True
+                if object.confidence > best_confidence:
+                    best_target = object
+                    best_confidence = object.confidence
+
+        # Create a negative response to our rotate service call
+        if best_target is None: 
+            response.response.data = "No %s object found" % target
+            return response
+
+        # Calculate how far we need to rotate
+        x_center = (best_target.x_min + best_target.x_max) / 2.0
+        angle_per_pixel = self.field_of_view / self.image_width
+        angle_to_rotate = (self.image_width/2.0 - x_center) * angle_per_pixel
+        rospy.loginfo("Center of target %f, rotation required %f degrees" % \
+                      (x_center, degrees(angle_to_rotate)))
+        # The angle needs to be in quaternion form (a 3D angle)
+        q = tf.transformations.quaternion_from_euler(0, 0, angle_to_rotate)
+
+        # Create goal and actionlib call to rotate
+        goal = move_base_msgs.msg.MoveBaseGoal()
+        goal.target_pose.header.frame_id = "base_link"
+        goal.target_pose.pose.orientation = Quaternion(*q)
+        self.move.send_goal(goal)
+        self.move.wait_for_result(rospy.Duration(60.0))
+
+        if self.move.get_state() == GoalStatus.SUCCEEDED:
+            response.response.data = "Rotated to %s" % target
+        else:
+            response.response.data = "Error rotating to %s" % target
+        return response
+
+    # Just sleep while the node is running
+    def run(self):
+        rate = rospy.Rate(1)
+        while (not rospy.is_shutdown() and self.ok):
```

Comment:
why not just rospy.spin() ?

Reply:
spin and then call `rospy.signal_shutdown("A String Reason for Humans")` when you want to exit. That should stop the spin.